File: Data Structures & Algorithms/binary-tree-diameter/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        self.max = 0

        def dfs(node, depth):
            leftHeight = depth
            rightHeight = depth

            if node.left:
                leftHeight = dfs(node.left, depth + 1)

            if node.right:
                rightHeight = dfs(node.right, depth + 1) 

            print(node.val, leftHeight, rightHeight, depth)

            self.max = max(self.max, leftHeight + rightHeight - 2*depth)

            return max(leftHeight, rightHeight)

        dfs(root, 0)

        return self.max
        

File: Data Structures & Algorithms/depth-of-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
private:
    int helper(TreeNode* node, int depth) {
        if (node == nullptr) return depth - 1;
        //if (node->left == nullptr && node->right == nullptr) return depth;
        return std::max(helper(node->left, depth+1), helper(node->right, depth+1));
    }
public:
    int maxDepth(TreeNode* root) {
        return helper(root, 1);
        
    }
};
